### PR TITLE
added logistic regression model and single-variant tests

### PR DESCRIPTION
**Pull request by jbloom22**
- support for wald, lr, and score tests
- added logreg doc
- added LogisticRegressionCommand
- added LogisticRegression
- added LogisticRegressionModel
- added LogisticRegressionSuite including Epacts comparison
- added LogisticRegressionModelSuite including R comparison
- added supporting resource files regressionLogistic*
- added chiSquaredTail function to Utils
- added chiSquaredTail teses to UtilsSuite

### CODE REVIEW

**Comment by jbloom22** [comment]
Timing on profile.vds (1KG, 2535 samples, 25956 variants) using three covariates (isFemale, PC1, PC2) with variant QC and export is 60-63s for wald and lrt, 10-12s for score, on one core locally (with ~2s for read).  Wald example:

```
hail: info: timing:
  read: 1.922s
  variantqc: 34.869ms
  annotatesamples table: 317.153ms
  annotatesamples fam: 60.988ms
  logreg: 783.904ms
  printschema: 2.034ms
  exportvariants: 1m2.3s
```

```
~/hail/build/install/hail/bin/hail \
--master local[1] \
read -i ~/data/profile.vds \
variantqc \
annotatesamples table -i ~/data/profile.ped -e IND_ID -t 'PC1: Double, PC2: Double' -r sa.pc \
annotatesamples fam -i ~/data/profile.fam \
logreg -y sa.fam.isCase -c sa.fam.isFemale,sa.pc.PC1,sa.pc.PC2 -t wald \
printschema \
exportvariants -c 'variant = v, beta = va.logreg.wald.beta, se = va.logreg.wald.se, zstat = va.logreg.wald.zstat, pval = va.logreg.wald.pval, nIter = va.logreg.fit.nIter, converged = va.logreg.fit.converged, exploded = va.logreg.fit.exploded, nNotCalled = va.qc.nNotCalled, nHomRef = va.qc.nHomRef, nHet = va.qc.nHet, nHomVar = va.qc.nHomVar' -o ~/data/profileHail/profile.sex.pc1.pc2.wald.tsv
```

More timing info and convergence analysis to come.

**Comment by jbloom22** [comment]
jenkins

**Comment by marivascruz** [comment]
Looks great. Any chance you can add Firth's logistic regression?

Thanks,

Manny

**Comment by jbloom22** [comment]
In fact I had Firth mixed into this branch but ripped it out when it was making the update too complicated. Whereas Wald, LRT, and score only require fitting the null model once, with the Firth likelihood ratio test you need to fit the null and full models per variant.

So plan is to add Firth, support for subsetting samples per variant (rather than imputing missing genotypes), and better tests by comparing Hail and R results for randomly generated datasets. I'd also like to add more [optional] user control on convergence criteria and on what's returned in annotations (for example, statistics for the other covariates...these are computed anyway). And there are ways to speed up the numerical linear algebra, this is a first pass.

Do you have thoughts on Firth LRT versus Wald? My understanding is that LRT is better calibrated for p-value, but would the Wald standard error for Firth be a useful annotation as well?

File: src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import org.kohsuke.args4j.{Option => Args4jOption}
+
+object LogisticRegressionCommand extends Command {
+
+  def name = "logreg"
+
+  def description = "Test each variant for association using the logistic regression model"
+
+  class Options extends BaseOptions {
+    // FIXME: specify that must be Boolean or 0 / 1
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala]: FIXME

File: src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    var ySA: String = _
+
+    @Args4jOption(required = false, name = "-c", aliases = Array("--covariates"), usage = "Covariate sample annotations, comma-separated")
+    var covSA: String = ""
+
+    @Args4jOption(required = false, name = "-t", aliases = Array("--test"), usage = "Statistical test: `wald', `lrt', or `score'")
+    var stat: String = "wald"
+
+    @Args4jOption(required = false, name = "-r", aliases = Array("--root"), usage = "Variant annotation root, a period-delimited path starting with `va'")
+    var root: String = "va.logreg"
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala]: Root is going away.  In the new model, you should have an expression in which the result of the test is available (say, through a variable called linreg) and the user can assign new annotations or compute whatever expressions they want.  See AnnotateVariantsTable `-c` option.

File: src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+      "s" -> (0, TSample),
+      "sa" -> (1, vds.saSignature))
+
+    val ec = EvalContext(symTab)
+    val sb = new StringBuilder
+    vds.saSignature.pretty(sb, 0, true)
+
+    val a = ec.a
+
+    def toDouble(t: BaseType, code: String): Any => Double = t match {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala]: Use `expr.DoubleNumericConversion.to`.

Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala]: You'll need to add Boolean support to it.  I'm fine with that.

Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala]: OK, that's not so straightforward with your error message.  I guess this is fine.

File: src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        .filter( (y, c, s) => y.isDefined && c.forall(_.isDefined))
+
+    val yArray = yForCompleteSamples.map(_.get).toArray
+    val y = DenseVector(yArray)
+
+    if (y.length == 0)
+      fatal("Every sample in this logistic regression is missing at least one covariate or phenotype")
+
+    if (! y.forall(yi => yi == 0d || yi == 1d))
+      fatal(s"For logistic regression, phenotype must be Boolean or numeric will all values equal to 0 or 1")
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala]: will => with.

File: src/main/scala/org/broadinstitute/hail/driver/VEP.scala [hunk 5]
```diff
@@ -36,7 +36,7 @@ object VEP extends Command {
 
   def name = "vep"
 
-  def description = "Annotation variants with VEP"
+  def description = "Annotate variants with VEP"
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/VEP.scala]: These small edits are awesome but should probably go in separate pull requests.

File: src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+import breeze.linalg._
+import org.apache.spark.rdd.RDD
+import org.broadinstitute.hail.Utils._
+import org.broadinstitute.hail.annotations.Annotation
+import org.broadinstitute.hail.stats.LogisticRegressionModel
+import org.broadinstitute.hail.variant._
+import org.broadinstitute.hail.expr._
+
+object LogisticRegression {
+  def name = "LogisticRegression"
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala]: Unused?

File: src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+  def buildX(gs: Iterable[Genotype], C: DenseMatrix[Double]): Option[DenseMatrix[Double]] = {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala]: Passing C into this function only to be concatenated seems like a strange choice.  I'd return an DenseVector and concatenate in the caller.

File: src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      val gtMean = gtSum.toDouble / nCalled
+      val gtArray = gs.map(_.gt.map(_.toDouble).getOrElse(gtMean)).toArray
+      Some(DenseMatrix.horzcat(new DenseMatrix(gtArray.length, 1, gtArray), C))
+    }
+  }
+
+  def apply(vds: VariantDataset, y: DenseVector[Double], cov: Option[DenseMatrix[Double]], stat: String): LogisticRegression = {
+    require(cov.forall(_.rows == y.length))
+    require(y.forall(yi => yi == 0 || yi == 1))
+    require{ val sumY = sum(y); sumY > 0 && sumY < y.size }
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala]: Is this how IntelliJ is indenting?  I'd expect a space before the opening curly brace.

File: src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala (hunk 6 above)
Comment (jbloom22) [inline src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala]: yes. I've also removed from here and LinearRegression.

File: src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala (hunk 8 above)
Comment (jbloom22) [inline src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala]: IntelliJ adds space and newlines:
```
    require {
      val sumY = sum(y); sumY > 0 && sumY < y.size
    }
```
I'll go with that.

File: src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala (hunk 7 above)
Comment (jbloom22) [inline src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala]: Changed name to buildGtColumn and returning a DenseMatrix for concatenation in caller.

File: src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala (hunk 2 above)
Comment (jbloom22) [inline src/main/scala/org/broadinstitute/hail/driver/LogisticRegressionCommand.scala]: How about I add this in a separate PR?

**Comment by jbloom22** [comment]
Ready for another look. I had to modify the classes some to make it work, particularly for getting the `type` out of the test. Now the type is with the Test rather than the TestResult, perhaps you see a better way? I'm waiting to squash until we've finalized.

Some related notes, mostly relevant topossible future PRs once we have a sense of performance:

I think LogisticRegressionNullFit should be a separate class, as it plays a conceptually and practically different role. I don't want to attach vectors of length nSamples (like mu) to each LogisticRegressionFit output, even though they would speed up the score test and first iteration of fitting per variant to not recompute them for every variant.  I did put some of this efficiency in the score test (only computing the extra coordinate of score and row / column of fisher per variant).  df would also then go away for LogisticRegressionFit, where it isn't used, but I'd add the diagonal of its inverse for use in Wald (see below). The model fit function would then take a LogisticRegressionNullFit to use to it's hearts content in the first iteration.

The bigger future gains will come from not computing or inverting the Fisher matrix at all in the iteration.

val sqrtW = sqrt(mu :* (1d - mu))
val QR = qr.reduced(X(::, *) :* sqrtW)
solve QR.R * deltaB = QR.Q.t * (y - mu) with R upper triangular (need to wrap lapack function)

for Wald: return diagonal of inverse as well, namely diagonal of inv(R)^T * inv(R), rather than inverting fisher again.

for Firth, modify score using:
val QQ = QR.Q :* QR.Q
val h = sum(QQ(*, ::))

for Score, a version of this may be faster:
val sqrtW = sqrt(mu :* (1d - mu))
val Qty0 = qr.reduced.justQ(X(::, *) :* sqrtW).t * ((y - mu) :/ sqrtW)
val chi2 = Qty0 dot Qty0

**Comment by jbloom22** [comment]
Updated runtime for LRT is about 50s:
```
~/hail/build/install/hail/bin/hail \
--master local[1] \
read -i ~/data/profile.vds \
annotatesamples table -i ~/data/profile.ped -e IND_ID -t 'PC1: Double, PC2: Double' -r sa.pc \
annotatesamples fam -i ~/data/profile.fam \
logreg -y sa.fam.isCase -c sa.fam.isFemale,sa.pc.PC1,sa.pc.PC2 -t lrt \
exportvariants -c 'variant = v, beta = va.logreg.lrt.beta, chi2 = va.logreg.lrt.chi2, pval = va.logreg.lrt.pval, nIter = va.logreg.fit.nIter, converged = va.logreg.fit.converged, exploded = va.logreg.fit.exploded' -o ~/data/profileHail/temp.profile.sex.pc1.pc2.lrt.tsv
```
```
  read: 2.193s
  annotatesamples table: 243.817ms
  annotatesamples fam: 22.080ms
  logreg: 367.098ms
  exportvariants: 50.473s
```

File: src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+object LogisticRegression {
+
+  def apply(vds: VariantDataset, y: DenseVector[Double], cov: Option[DenseMatrix[Double]], logRegTest: LogisticRegressionTest): LogisticRegression = {
+    require(cov.forall(_.rows == y.length))
+    require(y.forall(yi => yi == 0 || yi == 1))
+    require {val sumY = sum(y); sumY > 0 && sumY < y.size}
+
+    val n = y.length
+    val k = if (cov.isDefined) cov.get.cols else 0
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala]: No `.get`.  `cov.map(_.cols).getOrElse(0)`.

File: src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+          logRegTestBc.value.test(X, yBc.value, nullFitBc.value).toAnnotation
+        }
+      )
+    }
+
+    new LogisticRegression(rdd, logRegTest.`type`)
+  }
+
+  def buildGtColumn(gs: Iterable[Genotype]): Option[DenseMatrix[Double]] = {
+    val (nCalled, gtSum) = gs.flatMap(_.gt).foldRight((0, 0))((gt, acc) => (acc._1 + 1, acc._2 + gt))
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala]: `foldRight` is not tail recursive, and addition is associative.  You should use `foldLeft`.

File: src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+    // allHomRef || allHomVar || allNoCall || allHet
+    if (gtSum == 0 || gtSum == 2 * nCalled || nCalled == 0 || (gtSum == nCalled && gs.flatMap(_.gt).forall(_ == 1)))
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala]: This causes `gs` to be traversed twice (relatively expensive).  You might aggregate the last condition in the previous step.

File: src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      None
+    else {
+      val gtMean = gtSum.toDouble / nCalled
+      val gtArray = gs.map(_.gt.fold(gtMean)(_.toDouble)).toArray
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/methods/LogisticRegression.scala]: Again, map/getOrElse.

**Comment by jbloom22** [comment]
addressed comments, back to you.
